### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by majkinetor** [dnGrep/dnGrep#284]
Dngrep is very slow on larger number of files.

Perhaps multithread support would be good to have.

**Comment by doug24** [comment]
I assume you are using the latest version, 2.9.68.

Approximately how many files and directories are you searching?  
How large are the files - typical size and largest size?
Are they text files?
Are you searching in archives?

**Comment by JVimes** [comment]
@doug24 Does the verbose match count still slow things down? Can't remember where that ended up.

**Comment by doug24** [comment]
I think the latest version is pretty efficient now.  But it can't hurt to try:  Options - User Interface - uncheck 'Show verbose match count'.

**Comment by doug24** [comment]
tldr; it looks like we can get considerable performance improvement by changing the way the file filters are applied.  This is still a work in progress.

First, note that performance is highly dependent on the types of files you are searching, how many directories and files there are, what filters you are using, and if you are using Regex or Text for your search pattern.  It’s not difficult to create a Regex search text that takes a lot of time to process.  To keep things simple for this experiment I chose to use a text only search, and search many files of various sizes, but no very large files.

I did some performance profiling and found some unexpected results.  For a text search on a lot of files more time is spent finding the files than running the search.  The problem is many calls to Regex.IsMatch.
For a baseline, I’m using the current dnGrep x64 release 2.9.74.  My test directory is a set of old dev directories on my development laptop:  33,073 files in 3,647 folders, totaling 1.9 GB.  It’s a mixture of text and binary files.  I’m testing with two file pattern searches.  Here is the first:
![image](https://user-images.githubusercontent.com/17227248/37263509-c4ed5498-2565-11e8-806b-153e0263599f.png)

Here’s the result:
![image](https://user-images.githubusercontent.com/17227248/37263518-ce6b18c0-2565-11e8-8294-a2bee61da479.png)

The more file types you add to the search pattern, the worse it gets.  Adding *.csv to the files to search:
 
![image](https://user-images.githubusercontent.com/17227248/37263529-d457aa0a-2565-11e8-8899-19e8421a5d61.png)
![image](https://user-images.githubusercontent.com/17227248/37263536-d766327a-2565-11e8-86eb-f50a2fe5ab9d.png)
 
The profiler results show a big contributor to the processing time isn’t reading and searching within a file, it is just finding the files to search. And the cause of that?  A regular expression.  The current dnGrep uses regular expressions for the file filter.  It does not matter if you choose the “Regex” or “Asterisk pattern” option – the asterisk pattern is converted to a regular expression internally.  The application finds all the files in the directory tree (unfiltered) and starts evaluating all the filter parameters to every file.  The file pattern filer uses a regular expression to determine if the file should be included or not, and with more file types there are more regular expressions to be evaluated.
![image](https://user-images.githubusercontent.com/17227248/37263546-e2d391fc-2565-11e8-8e36-812f99e2a553.png)

First, I tied to make the regular expression matching more efficient but did not come up with anything helpful.
Next, I modified to code to so when the asterisk pattern option is selected, the pattern is used to filter files while the code is finding files in the directory tree.  This greatly reduces the number of files to process and is a lot more efficient. Repeating the two searches for *.xml and then for *.xml,*.csv give these results:  
![image](https://user-images.githubusercontent.com/17227248/37263558-ebc8ee74-2565-11e8-9649-9db48b65c1a2.png)
![image](https://user-images.githubusercontent.com/17227248/37263566-f03884c4-2565-11e8-980a-1408c68dd899.png)
 
Rerunning the profiler doesn’t show any more obvious bottlenecks.  The code to auto-detect the file encoding is near the top of the list now.  You can get performance improvement by preselecting the file encoding but may have reading errors if there are mix of encodings.

So now what if we run the file reading and text searching in parallel separate threads?  I limited it to 3 parallel threads on my 4 core CPU: running more threads in parallel isn’t significantly faster (probably due to contention for disk IO), and it’s not hard to peg the CPU at 100% lock the machine up.  A real implementation should consider the number of cores in the CPU at runtime.
 
![image](https://user-images.githubusercontent.com/17227248/37263644-4d66b332-2566-11e8-97ad-df28a22a5555.png)
![image](https://user-images.githubusercontent.com/17227248/37263648-520ad062-2566-11e8-94ba-cec9091bf320.png)
 
One down side with parallel processing is that you lose the ordering of the files by directory.  That may or may not matter to different users.  It is a problem for the application unit tests since the tests rely on the order of the results, and unit tests will fail if the files are processed in parallel.  Re-sorting the results after a run would be a lot of work to implement and would eat into the time saved.  Maybe parallel processing can be made a user configuration option?  I’m not sure it’s worth it to me, but for other search scenarios it may be more beneficial.
There is one breaking change with this change to the way the asterisk pattern is used.  dnGrep currently allows a pseudo regex to be input for an asterisk pattern.  There are two unit tests failing with these changes, and they test these two patterns as asterisk patterns:
t[a-z]st-file-*.*
t[ea]st-file-*.*
They are not valid patterns for a call to Directory.EnumerateFiles.  In the current version they are converted to the regular expressions, but I removed that path in the code.  I did investigate trying to determine if the pattern is usable as-is, and if not, fall back to the old behavior.  It may be possible, but more work than I’d want to put into it.  I would just replace these tests with others to test the types of things you can do with the [Windows supported patterns](https://msdn.microsoft.com/en-us/library/dd383571(v=vs.110).aspx#Anchor_2).

I think it’s reasonable to remove this feature for the performance benefit.  I suspect there are not many users creating patterns like this, and they can still create their own regular expression file filters by using the Regex option.

Another thing to note – the ‘paths to ignore’ filter is still implemented as regular expression matching and asterisk patterns are converted to regular expressions.  Using that option will still incur a performance hit – there isn’t any equivalent ignore directories/files as there is with the find files pattern filter.

**Comment by JVimes** [comment]
@doug24 Awesome research! Sorry if I'm not catching everything you said, but did you profile using the filesystem API's support for wildcards? I assume that should be fast, but maybe it's limited in a way I'm not thinking of.

Regarding the unit tests, I'd expect it to be looking for literal square brackets, dashes, etc. since it's "asterisk pattern", not "regex". Do what you like with those.

I'd say, let's speed up asterisk patterns by avoiding the regex and see if that's good enough for now.

**Comment by doug24** [comment]
@JVimes, yes, I profiled with the .net API support for wildcards, that was the most significant improvement in speed.  I think the filesystem support for wild cards is what most people would expect - * matches multiple characters, and ? matches one - patterns like *.txt, or test*.xml.  The wildcard characters can be used more than once in a pattern.   It also does special matching: when the pattern is *.xls, it will also also match .xlsx   And we still support multiple pattern strings separated by comma or semi-colon, like this  *.xml,*.json

**Comment by majkinetor** [comment]
Why not use everything search engine and get the files instantly ? It can be an option if it detects everything service running. Total Commander does this for example.

**Comment by JVimes** [comment]
@doug24 That sounds great. I think that's what everyone (myself included) expects the asterisk pattern to do.

@majkinetor mentioned I think this:
https://en.m.wikipedia.org/wiki/Everything_(software)

https://www.voidtools.com/faq/

It's an indexing search engine that can run as a service. Something to keep in mind in the future, if we add any value to what's already there.

**Comment by doug24** [comment]
Thanks for the suggestion - I had not thought about doing something like that.  I was looking at the links that John posted as well.  I have some reservations going that route.  I'd rather not depend on another installation, have to support different versions, or maybe even support different indexing applications.

From what I read they are reading the NTFS Master File Table.  I've found libraries that could do that directly without having to call another app, so I'd be more open to trying that. One problem is doing this bypasses the ACL and requires admin access.  

But here's the main thing:  the biggest problem isn't listing the files, it's filtering them. Unless I'm missing something, Everything isn't going to help with that - we can't pass our filter regex and other file filters to Everything - it just takes a simple string in it's search API.  So we'd have to ask it to return everything, and be back to the applying a regex to filter the files along with the date and size filters.

I re-ran my test above, just listing all the files with no filter, and it takes 4.5 seconds to find and iterate through 33,073 files.  If I add a file type filter (*.xml), it drops to just over 1 second.  So the built-in Windows directory and file listing is pretty good, and must be using some optimizations internally.

**Comment by majkinetor** [comment]
This is also very easy to do. I was using everything in various scripts since since 10 years ago. IPC via Windows API or `es.exe` - cli tool.

You can try it quickly out via chocolatey: `cinst everything --params /service`

**Comment by majkinetor** [comment]
Furthermore, this is true unix philosophy, right, do grep and let others do the search for files. 
Everything should be used as there currently exists nothing better on any OS. It doesn't only scan NTFS table but istalls itself as filter driver to update index realtime, unlike almost all other tools that require periodic indexing (most known x-plat is locate which requires updatedb).

**Comment by JVimes** [comment]
Doug's quick win is a good idea to get into the product right away. After that, it sounds like supporting Everything (if present on user machine) is worth researching, if it picks up on file changes in real-time (filter driver) and handles include/exclude patterns equivalent to ours. It'll require a contributing developer who is interested.

**Comment by majkinetor** [comment]
Everything supports more search options:

https://www.voidtools.com/support/everything/searching
https://www.voidtools.com/support/everything/searching/#regex

It makes sense to have an option to use everything search options exclusively, via another radio button besides 4 we already have now as it brings few new features to dnGrep (phrase count, recursion depth, etc.)

> It'll require a contributing developer who is interested.

I think this can be done within single day by someone who is already into the dnGrep code base.

Here is the sample
https://www.voidtools.com/support/everything/sdk/csharp/

**Comment by doug24** [comment]
Sorry, I'm in read-only mode during my day job (US-Pacific).  I will take a look at it when I get a chance, but don't be disappointed if I get the other changes in first.
